cg/meta/workflow/mutant/quality_controller/utils.py

```
@@ -0,0 +1,108 @@
+    for result in results:
+        if result.is_internal_negative_control:
+            internal_negative_control_result = result
+    return internal_negative_control_result.passes_qc
+
+
+def external_negative_control_qc_pass(results: list[SampleQualityResult]) -> bool:
+    for result in results:
+        if result.is_external_negative_control:
+            external_negative_control_result = result
+    return external_negative_control_result.passes_qc
+
+
+def get_sample_target_reads(sample: Sample) -> int:
+    return sample.application_version.application.target_reads
+
+
+def get_percent_reads_guaranteed(sample: Sample) -> int:
+    return sample.application_version.application.percent_reads_guaranteed
+
+
+def translate_samples_results(
+    raw_samples_results: SamplesResultsMetrics, samples_metadata: SamplesMetadataMetrics
+):
+    sample_name_to_id_mapping = {}
+    for sample_metadata in samples_metadata.samples.values():
+        sample_name_to_id_mapping[sample_metadata.sample_name] = sample_metadata.sample_internal_id
+
+    translated_samples_results = {}
+    for key, value in raw_samples_results.samples.items():
+        new_key: str = sample_name_to_id_mapping.get(key, key)
+        translated_samples_results[new_key] = value
+
+    samples_results = {"samples": translated_samples_results}
+
+    return SamplesResultsMetrics.model_validate(samples_results)
+
+
+def get_quality_metrics(
+    case_results_file_path: Path, case: Case, status_db: Store, lims: LimsAPI
+) -> QualityMetrics | None:
+    raw_samples_results: SamplesResultsMetrics = MetricsParser.parse_samples_results(
+        case_results_file_path
+    )
+
+    samples_metadata: SamplesMetadataMetrics = MetadataParser(
+        status_db=status_db, lims=lims
+    ).parse_metadata(case)
+
+    samples_results = translate_samples_results(raw_samples_results, samples_metadata)
+
+    if not samples_metadata:
+        return None
```

Could we raise an error here?

---

Yes!